src/com/sap/piper/cm/ChangeManagement.groovy

```diff
@@ -172,30 +172,141 @@ public class ChangeManagement implements Serializable {
+        // 2.2) the call in the narrower sense
+        //
+        // REVISIT: in case the customer uses a preconfigured image with the deploy tools already installed
+        //          there is no need for the npm call. --> in case the deployToolDependencies we should omit the
+        //          npm install call. In that case we should also drop the su and we don't launch the container as
+        //          root.
+
+        if (deployToolDependencies in List) {
+            deployToolDependencies = deployToolDependencies.join(' ')
+        }
 
-        if(rc != 0) {
-            throw new ChangeManagementException(
-                "Cannot upload file into transport request. Return code from  cm client: $rc.")
+        deployToolDependencies = deployToolDependencies.trim()
+
+        // --> Default config has been adjusted so that no dependencies needs to be installed.
+        // This is the case when an image is used which contains already all dependencies.
+        // In this case we don't invoke npm install and we run the image with the standard user
+        // already, since there is no need for being root. Hence we don't need to switch user also
+        // in the script.
+        boolean noInstall = deployToolDependencies.isEmpty()
+
+        def cmd = ['#!/bin/bash -e']
+
+        if (! noInstall) {
+            cmd << "npm install -g ${deployToolDependencies}"
+            cmd << "su ${osDeployUser}"
+        }
+        cmd << "fiori deploy -c \"${deployConfigFile}\" -- -y"
+
+        // 3.) execute the call in an appropirate docker container (node) and evaluate the return code
+        //     or let the AbortException bubble up.
+        script.withCredentials([script.usernamePassword(
+            credentialsId: credentialsId,
+            passwordVariable: 'password',
+            usernameVariable: 'username')]) {
+
+            // Set username and password for the fiori deploy call. The config file is configured to read the
+            // credentials from the environment (see above in the config file template).
+            // After installing the deploy toolset we switch the user. Since we do not su with option '-l' the
+            // environment variables are preserved.
+            def dockerEnvVars = docker.envVars ?: [:] + [ABAP_USER: script.username, ABAP_PASSWORD: script.password]
+
+            def dockerOptions = docker.options ?: []
+            if (!noInstall) {
+                // when we install globally we need to be root, after preparing that we can su node` in the bash script.
+                dockerOptions += ['-u', '0'] // should only be added if not already present.
+            }
+
+            script.dockerExecute(
+                script: script,
+                dockerImage: docker.image,
+                dockerOptions: dockerOptions,
+                dockerEnvVars: dockerEnvVars,
+                dockerPullImage: docker.pullImage) {
+
+                script.sh script: (cmd as Iterable).join('\n')
```

Comment:
Why `as Iterable`?

Reply:
In order to avoid deprecation warnings in the IDE.

Reply:
Resolved, the cmd can be defined above already as `Iterable`, with that we don't need to coerce here.